src/main/java/org/jenkinsci/remoting/engine/InboundTCPConnector.java

```diff
@@ -0,0 +1,267 @@
+import org.jenkinsci.remoting.protocol.IOHub;
+import org.jenkinsci.remoting.protocol.cert.DelegatingX509ExtendedTrustManager;
+import org.jenkinsci.remoting.protocol.cert.PublicKeyMatchingX509ExtendedTrustManager;
+import org.jenkinsci.remoting.protocol.impl.ConnectionRefusalException;
+import org.jenkinsci.remoting.util.KeyUtils;
+import org.jenkinsci.remoting.util.SSLUtils;
+
+/**
+ * Connects to a controller using inbound TCP.
+ */
+public class InboundTCPConnector extends AbstractEndpointConnector {
+    private static final Logger LOGGER = Logger.getLogger(InboundTCPConnector.class.getName());
+
+    private final JnlpEndpointResolver jnlpEndpointResolver;
+    private final List<URL> candidateUrls;
+    private final DelegatingX509ExtendedTrustManager agentTrustManager;
+    private final boolean keepAlive;
+
+    private URL url;
+    /**
+     * Name of the protocol that was used to successfully connect to the controller.
+     */
+    private String protocolName;
+
+    /**
+     * Tracks {@link Closeable} resources that need to be closed when this connector is closed.
+     */
+    @NonNull
+    private final List<Closeable> closeables = new ArrayList<>();
+
+    @Override
+    public URL getUrl() {
+        return url;
+    }
+
+    @SuppressFBWarnings(value = "HARD_CODE_PASSWORD", justification = "Password doesn't need to be protected.")
+    public InboundTCPConnector(
+            @NonNull String agentName,
+            @NonNull String secretKey,
+            @NonNull ExecutorService executor,
+            @NonNull EngineListenerSplitter events,
+            @NonNull Duration noReconnectAfter,
+            @CheckForNull List<X509Certificate> candidateCertificates,
+            boolean disableHttpsCertValidation,
+            @CheckForNull JarCache jarCache,
+            @CheckForNull String proxyCredentials,
+            @NonNull List<URL> candidateUrls,
+            @CheckForNull DelegatingX509ExtendedTrustManager agentTrustManager,
+            boolean keepAlive,
+            @NonNull JnlpEndpointResolver jnlpEndpointResolver) {
+        super(
+                agentName,
+                secretKey,
+                executor,
+                events,
+                noReconnectAfter,
+                candidateCertificates,
+                disableHttpsCertValidation,
+                jarCache,
+                proxyCredentials);
+        this.candidateUrls = new ArrayList<>(candidateUrls);
+        this.agentTrustManager = agentTrustManager;
+        this.keepAlive = keepAlive;
+        this.jnlpEndpointResolver = jnlpEndpointResolver;
+    }
+
+    private class EngineJnlpConnectionStateListener extends JnlpConnectionStateListener {
+
+        private final RSAPublicKey publicKey;
+        private final Map<String, String> headers;
+
+        public EngineJnlpConnectionStateListener(RSAPublicKey publicKey, Map<String, String> headers) {
+            this.publicKey = publicKey;
+            this.headers = headers;
+        }
```

Comment:
Could be a `record`.

Reply:
Too bad `JnlpConnectionStateListener` is an abstract class (for no good reason)